pkg/wallet/wallet.go

```diff
@@ -63,12 +68,71 @@ func (w *BaseWallet) DecryptMessage(encMessage []byte, toVerKey string) ([]byte,
 
 // PackMessage Pack a message for one or more recipients.
 func (w *BaseWallet) PackMessage(envelope *Envelope) ([]byte, error) {
-	return nil, fmt.Errorf("not implemented")
+	if envelope == nil {
+		return nil, errors.New("envelope argument is nil")
+	}
+	// get keypair from db
+	key, err := w.getKey(envelope.FromVerKey)
+	if err != nil {
+		return nil, fmt.Errorf("failed from getKey: %w", err)
+	}
+	// decode base58 public key
+	pubBytes := base58.Decode(key.Public)
+
+	// create 32 byte key
+	var pubKey [chacha.KeySize]byte
+	copy(pubKey[:], pubBytes)
+
+	var recipients []*[chacha.KeySize]byte
+	for _, verKey := range envelope.ToVerKeys {
+		// decode base58 ver key
+		verKeyBytes := base58.Decode(verKey)
+		// create 32 byte key
+		var recipient [chacha.KeySize]byte
+		copy(recipient[:], verKeyBytes)
+		recipients = append(recipients, &recipient)
+	}
+
+	// encrypt message
+	bytes, err := w.crypter.Encrypt(envelope.Message,
+		crypto.KeyPair{Priv: key.Private, Pub: &pubKey}, recipients)
+	if err != nil {
+		return nil, fmt.Errorf("failed from encrypt: %w", err)
+	}
+	return bytes, nil
 }
 
 // UnpackMessage Unpack a message.
 func (w *BaseWallet) UnpackMessage(encMessage []byte) (*Envelope, error) {
-	return nil, fmt.Errorf("not implemented")
+	var e authcrypt.Envelope
+	if err := json.Unmarshal(encMessage, &e); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal encMessage: %w", err)
+	}
+	var keysNotFound []string
+	for _, v := range e.Recipients {
+		recipVKeyB58 := v.Header.KID
+		// get keypair from db
+		key, err := w.getKey(recipVKeyB58)
+		if err != nil {
+			if errors.Is(err, ErrKeyNotFound) {
+				keysNotFound = append(keysNotFound, recipVKeyB58)
+				continue
+			}
+			return nil, fmt.Errorf("failed from getKey: %w", err)
+		}
+		// decode base58 public key
+		pubBytes := base58.Decode(key.Public)
+
+		// create 32 byte key
+		var pubKey [chacha.KeySize]byte
+		copy(pubKey[:], pubBytes)
+		bytes, err := w.crypter.Decrypt(encMessage, crypto.KeyPair{Priv: key.Private, Pub: &pubKey})
+		if err != nil {
+			return nil, fmt.Errorf("failed from decrypt: %w", err)
+		}
+		return &Envelope{Message: bytes}, nil
+	}
+	return nil, fmt.Errorf("no corresponding recipient key found in {%s}", keysNotFound)
```

Comment:
to be discussed, if we should be listing keys in error messages..

Reply:
I think it's ok to expose public keys now